### PR TITLE
[MM-48593] Create Prometheus SLOs per cluster

### PR DESCRIPTION
**Pull request by fmartingr**
#### Summary

Create the Sloth Prometheus Server Level object on cluster creation.

Removed a duplicated piece of code that wasn't being used: [kops_provisioner_installation_sli.go](https://github.com/mattermost/mattermost-cloud/blob/master/internal/provisioner/kops_provisioner_installation_sli.go) (logic is in [provisioner_installation_sli.go](https://github.com/mattermost/mattermost-cloud/blob/master/internal/provisioner/provisioner_installation_sli.go))

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-48593

#### Release Note

```release-note
Create Prometheus SLOs on cluster creation
```

### CODE REVIEW

File: internal/provisioner/cluster_provisioning.go [hunk 1]
```diff
@@ -342,6 +342,10 @@ func provisionCluster(
 		}
 	}
 
+	if err := createOrUpdateClusterSLOs(cluster, k8sClient, logger); err != nil {
+		return errors.Wrap(err, "failed to create cluster slos")
```
Comment (mirshahriar) [inline internal/provisioner/cluster_provisioning.go:346]:
```suggestion
		return errors.Wrap(err, "failed to create cluster SLOs")
```

File: internal/provisioner/sloth.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/mattermost/mattermost-cloud/k8s"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	slothv1 "github.com/slok/sloth/pkg/kubernetes/api/sloth/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func createPrometheusServiceLevel(psl slothv1.PrometheusServiceLevel, k8sClient *k8s.KubeClient, logger log.FieldLogger) error {
+	wait := 60
```
Comment (mirshahriar) [inline internal/provisioner/sloth.go:20]: Looks like we do not need this as a variable. Or may be `wait := time.Second * 60`. Your call. 0/5

File: internal/provisioner/sloth.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	err = updatePrometheusServiceLevel(psl, k8sClient, logger)
+	if err != nil {
+		return errors.Wrap(err, "failed to update cluster installation sli")
+	}
+	return nil
+}
+
+func deletePrometheusServiceLevel(psl slothv1.PrometheusServiceLevel, k8sClient *k8s.KubeClient, logger log.FieldLogger) error {
```
Comment (mirshahriar) [inline internal/provisioner/sloth.go:74]: This function never called.

**Review by mirshahriar — APPROVED** [review]
LGTM

File: internal/provisioner/sloth.go (hunk 3 above)
Comment (fmartingr) [reply internal/provisioner/sloth.go:74]: This method will be used in #777

File: internal/provisioner/sloth.go (hunk 2 above)
Comment (fmartingr) [reply internal/provisioner/sloth.go]: I've changed it to a default wait time using a duration instead of an int 👍

**Review by gabrieljackson — APPROVED** [review]
Very nice catch on the duplicated code 👍

File: internal/provisioner/kops_provisioner_installation_sli.go [hunk 4]
```diff
@@ -1,128 +0,0 @@
...
-	if err != nil && k8sErrors.IsNotFound(err) {
-		logger.Debugf("Sloth CRD doesn't exist on cluster: %s", err)
-		return nil
-	}
-	err = k8sClient.SlothClientsetV1.SlothV1().PrometheusServiceLevels("prometheus").Delete(ctx, sli, metav1.DeleteOptions{})
-	if err != nil {
-		return errors.Wrap(err, "failed to delete cluster installation sli")
-	}
-	return nil
-}
```
[the remark this answers is not in the document]
Comment (fmartingr) [reply internal/provisioner/kops_provisioner_installation_sli.go:128]: > Removed a duplicated piece of code that wasn't being used: [kops_provisioner_installation_sli.go](https://github.com/mattermost/mattermost-cloud/blob/master/internal/provisioner/kops_provisioner_installation_sli.go) (logic is in [provisioner_installation_sli.go](https://github.com/mattermost/mattermost-cloud/blob/master/internal/provisioner/provisioner_installation_sli.go))

**Pull request merged**